solr-search/src/main/java/edu/unc/lib/dl/search/solr/service/MultiSelectFacetListService.java

```diff
@@ -0,0 +1,140 @@
+                    searchRequest.isApplyCutoffs(), searchRequest.getAccessGroups());
+            if (selectedContainer == null) {
+                throw new NotFoundException("Invalid container selected");
+            }
+        }
+
+        // Turning off rollup because it is really slow
+        searchState.setRollup(false);
+
+        // Trim out retrieval of all facets which are also filtered, since they will be loaded separately
+        Set<String> facetFilterSet = searchState.getFacets().keySet();
+        if (facetFilterSet.size() > 0) {
+            List<String> toRetrieve = searchState.getFacetsToRetrieve().stream()
+                    .filter(f -> !facetFilterSet.contains(f)).collect(Collectors.toList());
+            searchState.setFacetsToRetrieve(toRetrieve);
+        }
+
+        // Calculate facets with facet filters applied
+        SearchRequest facetRequest = new SearchRequest(searchState, searchRequest.getAccessGroups(), true);
+
+        searchState.setRowsPerPage(0);
+        searchState.setResourceTypes(null);
+
+        SearchResultResponse resultResponse = searchService.getSearchResults(facetRequest);
+        resultResponse.setSelectedContainer(selectedContainer);
+        FacetFieldList resultFacets = resultResponse.getFacetFields();
+
+        // For each facet selected in the original search state, add facet list with results as
+        // if the that facet were not selected
+        for (Entry<String, List<SearchFacet>> facetEntry: searchRequest.getSearchState().getFacets().entrySet()) {
+            String facetName = facetEntry.getKey();
+            // Skip retrieval of facets that are filtered but not being retrieved in original request
+            if (!searchRequest.getSearchState().getFacetsToRetrieve().contains(facetName)) {
+                continue;
+            }
+
+            SearchState selectedState = (SearchState) searchState.clone();
+            selectedState.getFacets().remove(facetName);
+
+            selectedState.setFacetsToRetrieve(Arrays.asList(facetName));
+            SearchRequest selectedRequest = new SearchRequest(selectedState, searchRequest.getAccessGroups(), true);
+            SearchResultResponse selectedResponse = searchService.getSearchResults(selectedRequest);
+
+            // For MultivaluedHierarchicalFacet need to pull in the facets for child tiers
+            if (FacetFieldUtil.facetIsOfType(facetEntry.getValue(), MultivaluedHierarchicalFacet.class)) {
+                List<SearchFacet> merged = getMergedHierarchivalFacetValues(facetName, facetEntry.getValue(),
+                        selectedResponse, searchRequest);
+                resultFacets.add(new FacetFieldObject(facetName, merged));
+            } else {
+                resultFacets.add(selectedResponse.getFacetFields().get(0));
+            }
+        }
+
+        resultFacets.sort(searchRequest.getSearchState().getFacetsToRetrieve());
+        return resultResponse;
```

Comment:
How are the resultFacets being added to the resultReponse? Are they being updated in place from the FacetFieldList from line 77?

Reply:
`searchService.getSearchResults(facetRequest)` at line 75 performs a search and constructs a new result response, which contains a facetFields object populated with the facet values from that first search. I can add a comment that describes the steps here a little more. Effectively, it first issues a query for the search with all the facet filters applied, returning the facet fields which do not have filters applied to them (so if you were filtering by CONTENT_TYPE, the first query would just return PARENT_COLLECTION). Then it issues queries to get the facet values for each facet field which is being filtered, with all the OTHER filters applied but not its own filter (otherwise the only value it'd return would be the one being filtered on).